olp-cpp-sdk-dataservice-read/src/VersionedLayerClientImpl.cpp

```diff
@@ -0,0 +1,122 @@
+    : catalog_(std::move(catalog)),
+      layer_id_(std::move(layer_id)),
+      client_settings_(std::move(client_settings)),
+      pending_requests_(std::make_shared<PendingRequests>()) {
+  if (!client_settings_.cache) {
+    client_settings_.cache =
+        olp::client::OlpClientSettingsFactory::CreateDefaultCache({});
+  }
+}
+
+VersionedLayerClientImpl::~VersionedLayerClientImpl() {
+  pending_requests_->CancelPendingRequests();
+}
+
+olp::client::CancellationToken VersionedLayerClientImpl::GetData(
+    DataRequest data_request, Callback callback) {
+  auto context = olp::client::CancellationContext();
+  olp::client::CancellationToken token(
+      [context]() mutable { context.CancelOperation(); });
+
+  auto hrn = catalog_;
+  auto client_settings = client_settings_;
+  auto layer_id = layer_id_;
+  auto request_key = pending_requests_->GenerateRequestPlaceholder();
+  auto pending_requests = pending_requests_;
+  auto remove_request_and_callback = [pending_requests, request_key, callback](
+                                         CallbackResponse response) mutable {
+    pending_requests->Remove(request_key);
+    callback(std::move(response));
+  };
+
+  auto task = [context, remove_request_and_callback, hrn, data_request,
+               layer_id, client_settings]() mutable {
+    if (!data_request.GetDataHandle()) {
+      if (!data_request.GetVersion()) {
+        // get latest version of the layer if it wasn't set by the user
+        auto latest_version_response =
+            repository::CatalogRepository::GetLatestVersion(
+                hrn, context, data_request, client_settings);
+        if (!latest_version_response.IsSuccessful()) {
+          remove_request_and_callback(latest_version_response.GetError());
+          return;
+        }
+        data_request.WithVersion(
+            latest_version_response.GetResult().GetVersion());
+      }
+
+      // get data handle for a partition to be queried
+      auto partitions_response =
+          repository::PartitionsRepository::GetPartitionById(
+              hrn, layer_id, context, data_request, client_settings);
+      if (!partitions_response.IsSuccessful()) {
+        remove_request_and_callback(partitions_response.GetError());
+        return;
+      }
+      auto partitions = partitions_response.GetResult().GetPartitions();
+      if (partitions.empty()) {
+        remove_request_and_callback(client::ApiError(
+            client::ErrorCode::NotFound, "Partition not found"));
+        return;
+      }
+      data_request.WithDataHandle(partitions.front().GetDataHandle());
+    }
+
+    // finally get the data using a data handle
+    auto data_response = repository::DataRepository::GetBlobData(
+        hrn, layer_id, data_request, context, client_settings);
+    if (!data_response.IsSuccessful()) {
+      remove_request_and_callback(data_response.GetError());
+      return;
+    }
+    remove_request_and_callback(data_response.GetResult());
```

Comment:
I guess data_response can be passed to remove_request_and_callback directly.

Reply:
Unfortunately, no: inside remove_request_and_callback would has template parameter instead, specialized for each responses. And inside it would require to test the response once again against the error, which is an additional overhead.